provider/core/src/response.rs

```diff
@@ -318,10 +335,14 @@ where
             <M::Yokeable as Yokeable<'a>>::Output,
             PhantomData<&'a ()>,
         ) -> <M2::Yokeable as Yokeable<'a>>::Output,
+        M::Yokeable: Clone,
     {
-        DataPayload {
-            yoke: self.yoke.map_project(f),
-        }
+        DataPayload(match self.0 {
+            DataPayloadInner::Yoke(yoke) => DataPayloadInner::Yoke(yoke.map_project(f)),
+            DataPayloadInner::Ref(r) => {
+                DataPayloadInner::Yoke(Yoke::new_owned(r.clone()).map_project(f))
```

Comment:
issue: this does have a cost. I'm not sure how we can avoid this cost. However, there are a bunch of cases in the code where we don't _need_ owned mapping: ought we introduce `map_project_ref()` that does not require cloning carts, but does operate on references?

Reply:
I guess there are a few places in `datetime` where we map_project in runtime code, so I agree it would be nice if we didn't need to downgrade to ZeroFrom, but I'm not sure if that's avoidable in the general case?

Reply:
It's not avoidable in the general case, it is avoidable in DateTime.

Unfortunately it's not avoidable in the map property FFI code without significant changes :/ But we might be able to figure out a different way to type-erase there

Reply:
Currently we pay this cost on every `DataPayload` construction. With this PR, we only pay the cost when we `map_project`, so this is a strict improvement.

`map_project_cloned()` already operates on references.

Reply:
Yes, but it also clones the cart, which isn't great.

Something that operates on `self`, uses `&Y` and doesn't clone the cart is the missing niche here